myhoard/web_server.py

```diff
@@ -78,6 +79,59 @@ async def backup_list(self, _request):
                     response["backups"] = self.controller.state["backups"]
                 return json_response(response)
 
+    async def backup_preserve(self, request):
+        with self._handle_request(name="backup_preserve"):
+            stream_id = request.match_info["stream_id"]
+            body = await self._get_request_json(request)
+            preserve_until = body.get("preserve_until")
+
+            if preserve_until is not None:
+                try:
+                    preserve_until = datetime.fromisoformat(preserve_until).astimezone(timezone.utc)
+                    now = datetime.now(timezone.utc)
+                    if preserve_until < now:
+                        raise BadRequest("`preserve_until` must be a date in the future.")
+                except ValueError:
+                    raise BadRequest("`preserve_until` must be a valid isoformat datetime string.")
+
+            self.controller.mark_backup_preservation(stream_id=stream_id, preserve_until=preserve_until)
+            wait_for_applied_preservation = body.get("wait_for_applied_preservation")
+            if wait_for_applied_preservation:
+                self.log.info(
+                    "Waiting up to %.1f seconds for preservation of backup %s to be applied.",
+                    wait_for_applied_preservation,
+                    stream_id,
+                )
+                start = time.monotonic()
+                while True:
+                    backup = self.controller.get_backup_by_stream_id(stream_id)
+                    if not backup:
+                        # the backup was removed before preservation could be applied
+                        if preserve_until:
+                            return json_response({"success": False})
+                        # preservation was removed on time
+                        return json_response({"success": True})
+
+                    if (backup["preserve_until"] is None and preserve_until is None) or (
+                        backup["preserve_until"] == preserve_until.isoformat()
+                    ):
+                        self.log.info("Preservation for backup %s was applied.", stream_id)
+                        break
+
+                    elapsed = time.monotonic() - start
+                    if elapsed > wait_for_applied_preservation:
+                        self.log.info(
+                            "Preservation for backup %s was not applied up in %.1f seconds",
+                            stream_id,
+                            elapsed,
+                        )
+                        break
```

Comment:
I think this case should rather raise an error than return success, otherwise the timeout is rather meaningless as caller cannot anyway know what happened.

Reply:
should we raise the error also for backup_create? https://github.com/Aiven-Open/myhoard/blob/965421770f091ff1d3dc05ba4ddf3486a6fed0b9/myhoard/web_server.py#L55-L67  I was just following the current pattern

Reply:
Ah, I didn't realize it worked like this. I don't remember why it was done like that, seems weird looking at it now. I guess consistency is good but maybe then at least add some flag in the response to indicate the operation is still ongoing.